[숫자 야구 게임] 김홍동

nlom0218 opened the pull request.

nlom0218 committed:
docs: 숫자 야구 구현을 위한 기능 목록 작성하기

nlom0218 committed:
chore: 전체적인 흐름을 주석으로 작성하기

nlom0218 committed:
feat: 게임시작 문구를 출력하는 함수 만들기

MissionUtils 라이브러리에서 제공하는 Console API를 사용한다.

nlom0218 committed:
test: 게임시작 문구 출력 함수를 위한 테스트 작성

nlom0218 committed:
feat: 임의의 3자리 숫자를 선택하는 함수 만들기

MissionUtils 라이브러리에서 제공하는 Ramdom API를 사용한다.

nlom0218 committed:
test: 임의의 3자리 숫자 선택 함수를 위한 테스트 작성

0이 포함되어 있지 않은지 확인하기
3자리의 숫자인지 확인하기

nlom0218 committed:
feat: 숫자입력 문구를 출력하기

숫자 입력 문구는 게임 플레이어가 답변을 입력할 수 있어야 한다.
숫자 입력 문구는 답변을 인수로 전달하는 콜백함수를 호출한다.
게임 플레이어가 답변을 입력할 때까지 기다린다.
MissionUtils 라이브러리에서 제공하는 Console API를 사용한다.

nlom0218 committed:
feat: 게임 플레이어의 답변이 올바른지 검사하는 함수 만들기

3자리 숫자를 입력하지 않은 경우, 같은 숫자를 반복하여 입력한 경우, 0를 입력한 경우
그리고 숫자가 아닌 값을 입력한 경우 false를 반환한다.
그렇지 앟ㄴ은 경우 true를 반환한다.

nlom0218 committed:
test: 게임 플레이어의 답변이 올바른지 검사하는 함수를 위한 테스트 작성

nlom0218 committed:
feat: 플레이어의 답변이 올바른 형식이 아니라면 애플리케이션 강제 종료하기

throw문을 사용해 예외를 발생시킨후 애플리케이션을 종료한다.

nlom0218 committed:
feat: 게임 플레이어의 답변과 컴퓨터의 정답을 비교하여 결과를 반환하는 함수 만들기

낫싱: 컴퓨터가 선택한 3자리의 수가 하나도 없는 경우
볼: 컴퓨터가 선택한 3자리의 수와 같은 수가 다른 자리에 있는 경우
스트라이트: 컴퓨터가 선택한 3자리의 수와 같은 수가 같은 자리에 있는 경우
이때 볼이 스트라이크보다 앞에 나온다.
객체의 형태로 만들기(키: 볼 또는 스트라이크, 값: 키에 해당하는 결과의 개수)

nlom0218 committed:
test: 게임의 결과를 반환하는 함수를 위한 테스트 작성

nlom0218 committed:
feat: 게임결과 문구를 출력하는 함수 만들기

결과값이 담긴 객체를 매개변수로 받는다.
객체의 키에 해당하는 값을 바탕으로 출력 문구를 만든다.
모든 키의 값이 0이라면 낫싱을 출력하는 문구를 만든다.
MissionUtils 라이브러리에서 제공하는  API를 사용한다.

nlom0218 committed:
test: 게임결과 문구를 출력 함수를 위한 테스트 작성

콘솔에 출력하는 것에 대한 Jest 학습 후 다시 코드를 수정해야 함.

nlom0218 committed:
feat: 플레이어의 답변이 틀리면 다시 숫자입력 문구를 출력하기

3스트라이크가 아니라면 틀린 답변이다.
플레어이어의 답변이 맞을 때 까지 반복한다.

nlom0218 committed:
feat: 게임종료 문구를 출력하는 함수 만들기

게임 종료는 게임 결과가 3스트라이크일 때 실행된다.
게임 플레이어에게 질문을 던진다. 질문의 내용은 재시작할 것인지, 완전히 종료할 것인지이다.
구체적인 질문은 아래와 같다.
3개의 숫자를 모두 맞히셨습니다! 게임 종료 게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.
MissionUtils 라이브러리에서 제공하는 Console API를 사용한다.

nlom0218 committed:
feat: 게임 플레이어의 답변이 올바른지 검사하는 함수 만들기

1과 2를 제외한 답변은 올바른 형식이 아니다.

nlom0218 committed:
test: 게임 플레이어의 답변이 올바른지 검사하는 함수를 위한 테스트 작성

nlom0218 committed:
feat: 플레이어의 답변이 올바른 형식이 아니라면 애플리케이션 강제 종료하기

문을 사용해 예외를 발생시킨후 애플리케이션은 종료한다.

nlom0218 committed:
feat: 숫자 야구를 재시작하는 함수 만들기

숫자 야구가 재시작하는 조건은 다음과 같다.
게임 종료 후 1번을 선택한 경우

nlom0218 committed:
feat: 숫자 야구를 완전히 종료하는 함수 만들기

숫자 야구가 완전히 종료되는 조건은 다음과 같다.
게임 종료 후 2번을 선택한 경우
해당 함수는 강제로 종료하는 것이 아니기 때문에 예외를 발생시키지 않는다.
MissionUtils 라이브러리에서 제공하는 Console API를 사용한다.

nlom0218 committed:
docs: 기능 목록 추가

임의의 3자리 숫자를 선택하는 함수에 사용되는 메서드를 바꾸기 위해 기능 목록을 추가하였다.

nlom0218 committed:
feat: MissionUtils 라이브러리의 다른 메서드로 임의의 3자리 숫자를 선택하기

추가된 요구 사항을 유심히 보면 사용된 메서드가 다르다.
현재는 pickUniqueNumbersInRange()메서드를 사용한다.
이를 pickNumberInRange()메서드로 바꾸어 다시 구현한다.

nlom0218 committed:
docs: 게임결과 문구를 변경하기 위한 목록 추가

nlom0218 committed:
feat: 게임결과 문구가 3스트라이크라면 게임 종료 문구를 함께 출력하기

플레이어가 모든 숫자를 맞추면 3스트라이크가 출력된다.
이때 -3개의 숫자를 모두 맞히셨습니다! 게임 종료-도 함께 출력된다.

nlom0218 committed:
docs: 리팩토링 목록 추가

nlom0218 committed:
refactor: 함수(메서드) 이동하기

함수(메서드)의 성격에 맞게 이동한다.
App 클래스 전체적인 게임의 흐름을 담당한다.
game 객체는 게임 진행에 필요한 함수를 담당한다.
validation 객체는 유효성 검사에 필요한 함수를 담당한다.

nlom0218 committed:
refactor: 함수(메서드) 이동하기 #2

nlom0218 committed:
refactor: 변수명 수정하기

함수(메서드) 또는 변수가 하는 역할에 따라 변수명을 수정한다.
이름을 통해 의도를 나타내며 축약하여 작성하지 않도록 한다.

nlom0218 committed:
docs: 리팩토링 목록 추가

nlom0218 committed:
refactor: 함수 분리하기

게임을 재시작하는 함수를 분리한다.
src/App.js의 App 클래스에 reset() 메서드를 생성한다.

nlom0218 committed:
refactor: 변수명, 함수명 수정하기

App 클래스의 reset 메서드 명을 restart로 수정한다.

nlom0218 committed:
refactor: 매직 넘버 상수화하기

의미를 가지는 매직 넘버를 상수화한다.
const.js 파일을 만들어 상수화된 변수를 관리한다.

nlom0218 committed:
docs: 리팩토링 목록 추가

nlom0218 committed:
refactor: 변수명, 함수명 수정하기

const.js의 OPTION.END를 OPTION.EXIT로 수정한다.
App.js의 exit() 메서드와 이름을 동일하게 하기 위함이다.

nlom0218 committed:
refactor: if문 대신 배열 메서드 사용하기

playerInput를 검사하는 함수 내에서 if문 대신 array.every()을 사용하여 유효성을 검사한다.

nlom0218 committed:
refactor: 함수의 반환값 하나로 통일하기

game.quitWithException() 메서드는 에러를 생성한 후 종료를 시키기 때문에 반환하지 않는다.
App 클래스의 progress() 메서드와 end() 메서드는 App의 메서드를 반환한다.
game.printResult() 메서드는 아무런 값을 반환하지 않는다.

nlom0218 committed:
docs: 리팩토링 목록, 테스트 목록 추가하기

nlom0218 committed:
refactor: validation.option() else 제거하기

nlom0218 committed:
test: App class의 start() 테스트 코드 수정하기

nlom0218 committed:
test: game.getAnswer() 테스트 코드 수정 및 추가하기

nlom0218 committed:
test: game.getResult() 테스트 코드 수정 및 추가하기

nlom0218 committed:
test: game.printResult() 테스트 코드 수정하기

nlom0218 committed:
test: validation.playerInput() 테스트 코드 수정 및 추가하기

nlom0218 committed:
test: validation.optioin() 테스트 코드 수정하기

nlom0218 committed:
refactor: 변수명, 함수명 수정하기

const.js의 MESSAGE.INPUT을 MESSAGE.INPUT_NUMBER,
MESSAGE.END을 MESSAGE.INPUT_OPTION으로 수정한다. 이는 게임 플레이어의 input를 받기 때문에 수정을 한다.

nlom0218 committed:
refactor: 변수명, 함수명 수정하기 및 module.exports로 수정하기

의미와 의도가 들어난 이름으로 변수, 함수 이름 수정한다. answer이라는 단어는 누구의 answer인지 확실하게 하기 위해 computer, player를
앞에 붙여 의미를 명확히 한다. ball과 strike는 해당 개수를 뜻해야 하기 때문에 뒤에 count를 붙여 이름을 수정한다.
이에 따라 테스트 코드도 수정한다. 그리고 const.js의 객체를 exports하기 위해 module.exports를 사용한다. 각각 내보내는 것이 아니라
객체를 만들어 내보낸다.

nlom0218 committed:
refactor: 에러 메세지 추가하기

nlom0218 committed:
refactor: 에러 메세지 문구 수정하기

nlom0218 committed:
refactor: 피드백 리펙터링

nlom0218 committed:
docs: 리팩터링 추가

nlom0218 committed:
docs: 리팩터링 아이콘 추가

nlom0218 committed:
docs: 리팩터링 목록 추가

nlom0218 committed:
refactor: OutputView 클래스를 만들어 메시지 출력 관리하기

nlom0218 committed:
refactor: 사용자의 번호 입력 메시지 출력을 담당하는 메서드 만들기

nlom0218 committed:
refactor: 재시작, 종료 입력 메시지 출력을 담당하는 메서드 만들기

nlom0218 committed:
refactor: ComputerAnswer 클래스에서 컴퓨터의 정답 관련 필드와 메서드 만들기

setValue: 컴퓨터의 정답을 설정한다.
comparePlayerAnswer: 플레이어의 정답과 비교하여 볼과, 스트라이크의 개수를 반환한다.
resetValue: 컴퓨터의 정답을 재설정한다.

nlom0218 committed:
refactor: OutputView 클래스에서 게임의 결과 메시지 출력 메서드 만들기

기존 game 객체의 메서드에 존재했던 printResult 메서드를 OutputView 클래스로 옮긴다.

nlom0218 committed:
refator: Validator 클래스에서 유효성 검사 진행하기

nlom0218 committed:
refactor: 게임 종료를 담당하는 객체를 만들어 분리하기

nlom0218 committed:
refactor: Validator 클래스에서 유효성 검사 실패 시 앱 종료하기

nlom0218 committed:
refactor: GameManager 클래스를 생성하여 게임 진행을 담당하는 로직 분리하기

nlom0218 committed:
refactor: 플레이어의 입력에 따른 행동 분리하기

nlom0218 committed:
refactor: 컴퓨터의 정답을 숫자에서 문자열로 바꾸기

nlom0218 committed:
feat: 3스크라이크인지 판별하는 메서드 만들기

ComputerAnswer 클래스의 메서드로 만든다.

nlom0218 committed:
refactor: comparePlayerAnswer의 값을 바로 인자로 전달하기

OutputView의 printResult 메서드의 인자에 comparePlayerAnswer의 반환 값을 바로 넣는다.
받을 땐 구조분해 할당으로 ballCount와 strkeCount를 바로 받는다.

nlom0218 committed:
refactor: GameManager의 메서드 private으로 변경하기

nlom0218 committed:
refactor: Validator를 클래스가 아니라 일반 객체로 바꾸기

nlom0218 committed:
refactor: 상수 관리하기

mayreel0 commented on src/Model/ComputerAnswer.js:
```diff
@@ -0,0 +1,43 @@
+    while (numberArr.length < 3) {
+      const number = Random.pickNumberInRange(1, 9);
+      if (!numberArr.includes(number)) numberArr.push(number);
+    }
```
중복 값을 배제하기 위해서는 의미상 Set 객체를 생성하는 것도 좋은 방법이라고 생각합니다. 
https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Set

nlom0218 replied:
네! Set 객체를 이용해봐야겠어요. 감사합니다👍

mayreel0 commented on src/Model/ComputerAnswer.js:
```diff
@@ -0,0 +1,43 @@
+    this.#setValue();
+  }
+
+  #setValue() {
```
보통 setter를 외부에서 필드의 값을 변경할 때 많이 사용하는데 해당 메서드는 생성에 가까운 것 같아요. create나 initialize같은 이름도 고려해보는게 좋을 것 같네요.

nlom0218 replied:
그러쿤요! 말씀해주신 내용을 참고하여 메서드명을 바꾸어야 겠습니다ㅎㅎ

mayreel0 commented on src/Model/ComputerAnswer.js:
```diff
@@ -0,0 +1,43 @@
+    let strikeCount = 0;
+
+    [...playerAnswer].forEach((number, idx) => {
+      if (![...this.#value].includes(number)) return;
```
문자열 상태에서도 includes가 가능한걸로 알고 있는데 배열로 바꿔주신 이유가 있을까요??

nlom0218 replied:
오.. 몰랐던 지식을 얻고 갑니다! 감사해요😄

mayreel0 commented on src/Model/Validator.js:
```diff
@@ -0,0 +1,42 @@
+  playerOption(option) {
+    let isValid = false;
+
+    if (option === OPTION.restart || option === OPTION.exit) isValid = true;
```
isValid 변수를 선언할 필요 없이 early return 했어도 좋았을 것 같아요!
```js
playerOption(option) {
  if (option === OPTION.restart || option === OPTION.exit) { 
    return;
  }

  Quit.withException(ERROR_MESSAGE.option);
}

```

nlom0218 replied:
오.. 말씀해주신 코드로 리팩터링을 진행해봐야겠어요
필요없는 변수를 굳이 선언을 했었네요ㅠㅠ😭

mayreel0 commented on src/Model/Validator.js:
```diff
@@ -0,0 +1,42 @@
+    if (playerAnswer.length !== 3) Quit.withException(ERROR_MESSAGE.length);
+  },
+
+  checkNotIncludesZero(playerAnswer) {
```
메서드 이름은 0을 포함하지 않는 의미인데 동작은 0을 포함하는 동작이네요 ?_?

nlom0218 replied:
메서드 이름을 얼른 바꾸러 가야겠어요ㅋㅋㅋ
세심한 리뷰 감사합니다!!

mayreel0 commented on src/Model/Validator.js:
```diff
@@ -0,0 +1,42 @@
+  },
+
+  checkNumber(playerAnswer) {
+    if (isNaN(playerAnswer)) Quit.withException(ERROR_MESSAGE.isNotNumber);
```
`isNaN()` 함수는 어떤 값이 NaN인지 판별하는데
인수가 Number 형이 아닌 경우, 그 값은 먼저 숫자로 형 변환을 합니다.
형 변환에 의존하면 신뢰도가 떨어지고 예기치 못한 결과를 낼 수도 있지요. 
반면에 `Number.isNaN()`를 사용하면 강제로 매개변수를 숫자로 변환하지 않고, 
주어진 값의 유형이 Number이고 값이 NaN이어야만 true를 반환합니다.

nlom0218 replied:
`Number.isNaN()` 또 새로운 지식을 얻습니다!!

정환님 덕분에 많은 배움을 얻습니다! 정말 감사해요😁
그냥 놓치고 갈 수 있었던 부분을 세심하게 알려주셔서 리뷰를 읽는 것 만으로도 많은 배움이 있었습니다👍👍

hwanfront reviewed (commented):
LGTM! 그동안에 고생한 흔적이 보이네요. 테스트코드도 잘 짜신 것 같아요. 2주차 피드백이었나? 영상 자료가 있는 피드백이 있었을거에요. 그거 한번 시청해보시면 객체를 잡는데에 또 새로우실거에요. 한번 시청해보시는것도??

nlom0218 committed:
refactor: ComputerAnswer의 setValue 메서드 수정

메서드 이름을 createValue로 numberArr 변수를 numbers로 변경한다.
Set 객체를 활용하여 중복된 숫자를 배제한다.

nlom0218 committed:
refactor: ComputerAnswer의 comparePlayerAnswer includes 메서드 사용

문자열에서도 includes메서드 사용이 가능하기 때문에 문자열을 배열로 바꾸는 과정을 생략한다.

nlom0218 committed:
refactor: Validator의 playerOption 메서드에서 isValid 변수 제거

early return으로 인해 isValid 변수가 필요없다.

nlom0218 committed:
refactor: Validator의 checkNotIncludesZero 메서드 이름 수정

checkIncludesZero로 메서드 이름을 변경한다. 메서드 이름과 동작을 일치시킨다.

nlom0218 committed:
refactor: Validator의 checkNumber 메서드의 isNaN을 Number.isNaN로 변경

youzysu commented on src/Model/Validator.js:
```diff
@@ -0,0 +1,41 @@
+  },
+
+  checkNotOverlap(playerAnswer) {
+    if ([...new Set([...(playerAnswer + '')])].length !== 3)
```
여기서 배열로 안바꾸고 set.size를 사용하셔도 될 것 같습니다!

nlom0218 replied:
Set 객체를 자주 사용하지 않아 메서드들을 깜빡하고 있는데, 이렇게 짚어 주셔서 감사해요!👍

youzysu commented on src/Model/ComputerAnswer.js:
```diff
@@ -0,0 +1,44 @@
+      numbers = [...new Set(numbers)];
+    }
+
+    this.#value = numbers.join('');
```
혹시 string 타입으로 숫자를 비교하신 이유가 무엇일까요!?
스프레드 연산자가 반복적으로 중첩해서 쓰이고 있어서 value를 배열 그대로, playerAnswer를 array로 바꿔서 인자로 전달하는 방법도 좋을 것 같아요!

nlom0218 replied:
제가 문자열로 저장한 까닭은 3스트라이크인지 판단을 바로 할 수 있기 때문이에요.

아래에 등장하는 `isThreeStrike` 메서드에선 컴퓨터의 정답과 플레이어의 정답을 바로 비교할 수 있게끔 문자열로 저장했어요. 하지만 배열로 저장한 후 이후에 문자열로 바꾸어 비교해도 좋을 거 같네요!

youzysu reviewed (commented):
클래스 분리가 잘되어 있어서 이해하기 쉬웠습니다 ㅎㅎ
코드 리뷰 남겨두었어요! 작은 의견이지만 도움이 되면 좋겠어요! 수고하셨습니다 👍

quack777 commented on __tests__/GameTest.js:
```diff
@@ -0,0 +1,120 @@
+  });
+});
+
+describe.each([
+  [438, "421", { ballCount: 0, strikeCount: 1 }],
+  [438, "253", { ballCount: 1, strikeCount: 0 }],
+  [438, "328", { ballCount: 1, strikeCount: 1 }],
+  [438, "843", { ballCount: 3, strikeCount: 0 }],
+  [438, "179", { ballCount: 0, strikeCount: 0 }],
+  [769, "769", { ballCount: 0, strikeCount: 3 }],
+])("game.getResult(%d, %s) 테스트", (answer, input, expected) => {
+  test(`${expected}을 반환하는 테스트`, () => {
+    expect(game.getResult(answer, input)).toEqual(expected);
+  });
+
+  test("객체의 key로 ball과 strike를 가지는지 테스트", () => {
+    const objectShape = {
+      ballCount: expect.any(Number),
+      strikeCount: expect.any(Number),
+    };
+
+    expect(game.getResult(answer, input)).toEqual(
+      expect.objectContaining(objectShape)
+    );
+  });
+});
```
`describe.each` 잘 사용하신 것 같네요! 저도 한 번 공부해봐야겠어요👍👍

quack777 commented on src/Controller/GameManager.js:
```diff
@@ -0,0 +1,58 @@
+  #computerAnswer;
+
+  constructor(compuerAnswer) {
+    this.#computerAnswer = compuerAnswer;
```
```suggestion
    this.#computerAnswer = new ComputerAnswer();
```
constructor에서 copuerAnswer을 받아 오지 않고 이렇게 수정해도 괜찮을 것 같아요~
copuerAnswer 오타 있는 것 같네요!

nlom0218 replied:
오타가 있었네요!ㅋㅋㅋ 얼른 수정해야겠어요~
저도 처음엔 `GameManager` 클래스의 생성자 함수에서 은총님 말씀처럼 코드를 작성했는데, 어떤 것이 더 알맞는 방법인지 몰라서 이것저것 고민을 했었어요!

nlom0218 committed:
refactor: createValue에서의 Set 객체 사용하여 컴퓨터의 정답 정하기

quack777 commented on src/Controller/GameManager.js:
```diff
@@ -0,0 +1,58 @@
+  #actionAboutPlayerAnswer(playerAnswer) {
+    if (this.#computerAnswer.isThreeStrike(playerAnswer)) return this.#end();
+
+    return this.#play();
```
`return`이 불필요하다고 느껴지는데 제거해도 좋을 것 같습니다😉

quack777 replied:
아래 49번째 라인 `retrun`도 마찬가지입니다

nlom0218 replied:
아하..! 흐름을 보니 필요가 없어 보이네요! 감사합니다~

nlom0218 committed:
checkNotOverlap에서의 Set.size 메서드 사용하기

quack777 commented on src/Model/ComputerAnswer.js:
```diff
@@ -0,0 +1,44 @@
+      const num = Random.pickNumberInRange(1, 9);
+      numbers.push(num);
+      numbers = [...new Set(numbers)];
+    }
```
```suggestion
    const numbers = new Set();
    while (numbers.size < 3) {
      const num = Random.pickNumberInRange(1, 9);
      numbers.add(num);
    }
    }
```
`Set`을 사용하셨기 때문에 배열 대신 전부 `Set`을 활용하는 것은 어떨까요??

nlom0218 replied:
네, 맞아요! Set를 활용하여 리팩터링을 다시 했습니다ㅎㅎ 감사합니다!

quack777 commented on src/Model/ComputerAnswer.js:
```diff
@@ -0,0 +1,43 @@
+    let ballCount = 0;
+    let strikeCount = 0;
+
+    [...playerAnswer].forEach((number, idx) => {
```
`[...playerAnswer]` 정말 좋은 것 같습니다👍👍

shackstack replied:
이렇게도 쓸 수 있군요! 한수 배워갑니다!

quack777 reviewed (commented):
홍동님 리팩토링하느라 고생많으셨습니다!! 테스트 코트도 정말 잘 짜시고 배울 점이 많은 코드였습니다👍👍👍

nlom0218 committed:
refactor: GameManager의 constructor수정 및 불필요한 retrun 제거하기

Rki0 commented on src/View/OutputView.js:
```diff
@@ -0,0 +1,29 @@
+
+    this.#printBallAndStrikeCount(ballCount, strikeCount);
+
+    if (strikeCount === 3) Console.print(MESSAGE.success);
```
`strikeCount`의 3은 정답을 의미하는 숫자니까 상수화하는게 좋을  것 같습니다!

nlom0218 replied:
그러쿤요! 미쳐 생각지 못했던 내용이이네요. 감사합니다👍

Rki0 commented on src/Model/ComputerAnswer.js:
```diff
@@ -0,0 +1,43 @@
+    let strikeCount = 0;
+
+    [...playerAnswer].forEach((number, idx) => {
+      if (!this.#value.includes(number)) return;
```
저는 볼, 스트라이크 조건에 이 조건을 다 넣어줬는데, 처음부터 `return` 처리를 해도 됐겠네요😂
훨씬 깔끔한 방법인 것 같습니다!

Rki0 commented on src/Controller/GameManager.js:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  #play() {
+    InputView.readPlayerAnswer((playerAnswer) => {
```
콜백 함수를 완전히 분리해서 하나의 메서드로 만드시면 훨씬 깔끔할 것 같습니다!

nlom0218 replied:
네, 저도 고민을 했습니다. 많은 분들이 콜백함수를 하나의 메서드로 만들어서 분리하셔서 저도 도전해볼까 했는데, 이번 기회에 도전해봐야겠습니다ㅎㅎ

Rki0 commented on src/Model/ComputerAnswer.js:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  #createValue() {
+    let numbers = new Set();
```
`Set` 객체로 값을 생성하고 배열로 변환하는 과정이 많이 보이는 것 같습니다.
처음 값을 생성할 때 배열로 만들고, 중복값 검증에서만 `Set`을 사용하시는 방법은 어떻게 생각하시는지 궁금합니다!

nlom0218 replied:
말씀해주신 리뷰 내용을 이전 코드에서 적용을 했었습니다.
하지만 지금과 같은 코드로 바꾼 이유는 `set` 객체에는 동일한 값을 저장하지 않는 특징이 있어, 굳이 포함되었는지를 확인하지 않아도 되기 때문에 오히려 `set` 객체를 처음부터 사용하는 것이 더욱 깔끔해 보였습니다. 다만 말씀해주시면 내용처럼 변환하는 과정이 많이 보이는 것이 있죠ㅠㅠ 그래서 많이 고민입니다.

좀 더 고민을 해봐야하는 부분이 생겼습니다! 감사합니다.😁

Rki0 commented on src/libs/const.js:
```diff
@@ -0,0 +1,32 @@
+const MESSAGE = {
+  start: '숫자 야구 게임을 시작합니다.',
```
저는 상수라는 점을 더 명확하게 하기 위해 대문자와 _(밑줄)을 사용합니다. 
홍동님께서는 상수를 선언하실 때 소문자를 사용하신 이유가 있으신가요?

nlom0218 replied:
Aribnb 코딩 컨벤션을 보면 어떤 객체가(위에선 `MESSAGE`) 상수를 관리하는 객체라는 것을 의미한다면 그 객체의 값은 굳이 대문자로 하지 않아도 된다고 합니다.
https://github.com/parksb/javascript-style-guide#naming--uppercase

기영님의 리뷰는 객체로 관리하는 것이 아니라 하나의 상수로 관리하실 때를 말씀하시는 건가요? 저도 객체로 관리하지 않고 하나하나 상수로 관리하면 `MESSAGE_START`와 같이 관리합니다ㅎㅎ 하지만 만약 여러 메시지가 있다면 메시지라는 공통점이 있기 때문에 `MESSAGE`를 객체로 만들고 관리를 합니다!

Rki0 reviewed (commented):
MVC를 명확하게 나누신 것이 인상적입니다.
`readLine`에 들어갈 콜백 함수를 분리해서 메서드로 만들면 더 깔끔할 것 같습니다 :)

nlom0218 committed:
refactor: 12월 29일 코드 리뷰에 대한 리팩터링

1. 숫자 3을 상수화한다.
2. readLine에 들어갈 콜백 함수를 분리하여 메서드를 만든다.

sangpok commented on __tests__/ValidationTest.js:
```diff
@@ -0,0 +1,55 @@
+const validation = require("../src/libs/validation");
+
+describe("validation.checkPlayerAnswer() 테스트", () => {
+  test.each([["123"], ["456"], ["258"]])(
+    "validation.checkPlayerAnswer(%s) => true",
+    (palyerAnswer) => {
+      expect(validation.checkPlayerAnswer(palyerAnswer)).toBeTruthy();
+    }
+  );
```
Jest 메소드를 잘 활용하시는 게 인상적입니다! 저는 익숙하지 않아서 일일이 다 쳤는데 이런 방법이 있었군요👍 배워갑니다ㅎㅎ

nlom0218 replied:
Jest 공식문서에 다양한 메서드가 있더라고요! 참고하면서 테스트 코드 작성하시면 도움이 많이 되실거에요!

sangpok commented on docs/README.md:
```diff
@@ -0,0 +1,206 @@
+    - 3자리 숫자를 입력하지 않은 경우, 같은 숫자를 반복하여 입력한 경우, 0를 입력한 경우 `false`를 반환한다.
+    - 또한 숫자가 아닌 값을 입력하여도 `false`를 반환한다.
+    - 그렇지 않은 경우 `true`를 반환한다.
+  - [x] test: 게임 플레이어의 답변이 올바른지 검사하는 함수를 위한 테스트 작성
```
기능 요구 사항을 만들 때 이렇게 테스트할 것도 적어두면 흐름을 정리하기 좋아보이네요 !! 배워갑니다!

nlom0218 replied:
저는 기능목록에 기능에 필요한 목록 뿐 아니라 테스트, 리팩터링을 위한 것도 정리하려고 노력해요!ㅎㅎ

sangpok commented on src/Controller/GameManager.js:
```diff
@@ -0,0 +1,63 @@
+const Quit = require('../libs/Quit');
+const { MESSAGE, OPTION } = require('../libs/const');
+
+class GameManager {
```
게임을 진행하는 GameManager는 App이 실행된 이후부턴 인스턴스가 재생성되거나 다른 인스턴스가 만들어지지 않아보여요. 이런 경우라면 싱글톤을 유지하는 클래스라든지, 또는 객체로 선언하는 것이 좋을 것 같은데, 따로 클래스로 정의하신 이유가 있으실까요?!

nlom0218 replied:
일단 `GameManager` 라는 클래스를 만들어 모델과 뷰를 연결시키고 싶었어요. 또한, 게임의 전체적인 흐름을 담당하는 역할을 `GameManager` 클래스가 맡게되는 것이고요. 

사실 이번 프리코스 이전에는 `class`을 이렇게 적극적으로 사용해본 경험이 없어요..ㅠㅠ 그래서 무엇이 옳다, 이때는 객체가 좋다. 라는 기준에 대해서 혼란을 겪고있는 상황입니다. 말씀해주신 내용을 바탕으로 싱글톤을 유지하는 클래스에 대해 학습을 해야겠습니다!ㅎㅎ

sangpok commented on src/Controller/GameManager.js:
```diff
@@ -0,0 +1,63 @@
+    this.#actionAboutPlayerAnswer(playerAnswer);
+  }
+
+  #actionAboutPlayerAnswer(playerAnswer) {
+    if (this.#computerAnswer.isThreeStrike(playerAnswer)) this.#end();
+
+    this.#play();
+  }
```
이런 로직이라면 3스트라이크일 경우 this.#end()로 갔다가 비동기인 InputView.readCommand()를 만나면 다시 넘어와서 this.#play()를 호출하지 않나요? 지금은 Console.readLine의 자체적인 콜백 특성때문에 아웃풋상으론 옳게 동작하게 보이겠지만, 사실상 두 번 호출이 되는 거라 요구사항이 원하는 방향과는 다른 방향일 것이라는 생각이 듭니다😅 테스트 해보시려면 this.#play()위에 콘솔 한번 찍어보시면 될 듯 합니다!

nlom0218 replied:
오호! `this.#end()` 메서드 앞에 `return`을 해야겠어요! 감사합니다ㅎㅎ

sangpok commented on src/Controller/GameManager.js:
```diff
@@ -0,0 +1,63 @@
+  }
+
+  #play() {
+    InputView.readPlayerAnswer(this.#handlePlayerAnswer.bind(this));
```
bind를 쓰지 않는 방법으론 Arrow함수로 인자를 넘겨주는 방법도 있습니다 :)
```javascript
InputView.readPlayerAnswer((param) => this.#handlePlayerAnswer(param));
```

bind를 하게 되면 새로운 함수를 만들어서 리턴하는 것과 같아서, 메모리 낭비라는 이야기를 들은 것 같아서요. 위처럼 진입함수를 만들어주면 원래의 함수를 쓸 수 있기에 낭비를 피할 수 있습니다. 물론 진입 비용이 있긴 하지만, 영향을 줄 정도의 큰 비용은 아니라고 알고 있습니다 👍

nlom0218 replied:
메모리 낭비가 있을 수 있군요! 저도 처음엔 `arrow` 함수를 만들어 인자로 넘겼었는데, 어떤 방법이 더 좋고 효율적인지 고민하다 `bind()`라는 메서드를 공부하고 한 번 사용해봤습니다!ㅎㅎ

sangpok commented on src/Model/ComputerAnswer.js:
```diff
@@ -0,0 +1,43 @@
+      numbers.add(num);
+    }
+
+    this.#value = [...numbers].join('');
```
요 부분을 보니 `#value`의 타입이 `string`이군요! 저는 막연히 `Array<number>`로 변환해서 값을 저장해야 한다고 생각했는데, 곰곰이 생각해보니 굳이 변환하지 않아도 될 것 같네요! 좋은 포인트인 것 같습니다👍

sangpok commented on src/Controller/GameManager.js:
```diff
@@ -0,0 +1,63 @@
+    return Quit.game();
+  }
+
+  #restart() {
+    this.#computerAnswer.resetValue();
+    this.#play();
+  }
```
앗 저는 computerAnswer를 Class로 만드신 이유가 재할당 때문이라고 생각했는데, 이렇게 resetValue()가 따로 있으면 객체에 좀 더 가까운 활용같다는 생각이 들어요! 아래와 같이 쓰시면 클래스를 좀 더 잘 활용하는 방법이 될 것 같습니다 👍
```suggestion
  #restart() {
    this.#computerAnswer = new ComputerAnswer(); // 인스턴스 재생성
    this.#play();
  }
```

nlom0218 replied:
좋은 지적 감사합니다! 아직 객체, 클래스에 대한 명확한 구분이 어려워 헤매고 있는 중입니다ㅠㅠ😭 객체와 클래스, 언제 무엇을 사용하는지에 대한 기준을 찾아보며 공부를 해야겠어요ㅎㅎ

sangpok commented on src/View/InputView.js:
```diff
@@ -0,0 +1,14 @@
+const { MESSAGE } = require('../libs/const');
+const { Console } = require('@woowacourse/mission-utils');
+
+class InputView {
+  static readPlayerAnswer(callback) {
+    Console.readLine(MESSAGE.inputNumber, callback);
+  }
+
+  static readCommand(callback) {
+    Console.readLine(MESSAGE.inputOption, callback);
+  }
+}
```
InputView는 Class로 선언됐지만 데이터를 다루지 않는 클래스, 다시 말해 필드가 없이 메소드만 제공하는 클래스이므로 여기에 정적 메소드를 활용하는 것보다는 객체로서 활용하는 것이 더 좋아보여요👍

```suggestion
const InputView = {
  readPlayerAnswer(callback) {
    Console.readLine(MESSAGE.inputNumber, callback);
  }

  readCommand(callback) {
    Console.readLine(MESSAGE.inputOption, callback);
  }
}
```

Static Method를 활용할 때에는 특정 클래스 인스턴스가 아니라 모든 클래스 인스턴스에 필요할 때 유용하게 쓰일 수 있습니다!

nlom0218 replied:
네! 맞습니다ㅎㅎ `InputView`, `OutputView`가 굳이 클래스여야 하나? 라는 의문이 들더라고요! 객체로 활용해야 겠습니다ㅎㅎ

sangpok commented on src/View/OutputView.js:
```diff
@@ -0,0 +1,29 @@
+  }
+
+  static printResult({ ballCount, strikeCount }) {
+    if (ballCount === 0 && strikeCount === 0) {
```
궁금한 게 있는데요! [Airbnb 코드 컨벤션 / Comparison Operators & Equality](https://github.com/airbnb/javascript#comparison-operators--equality) 에 따르면 boolean 타입을 체크할 때면 shortcut을 사용하라고 했는데, 예외적으로 문자열과 숫자에 대한 비교는 명시적으로 하라고 되어있습니다. 그렇다면 이렇게 Count를 비교할 땐 명시적으로 0과 비교를 해주는 게 코드 컨벤션에 따르는 건지 문득 궁금해졌습니다! ㅎㅎ

sangpok commented on src/View/OutputView.js:
```diff
@@ -0,0 +1,29 @@
+
+    this.#printBallAndStrikeCount(ballCount, strikeCount);
+
+    if (strikeCount === STRIKE_COUNT) Console.print(MESSAGE.success);
```
저에겐 요 코드는 스트라이크 카운트랑 스트라이크 카운트랑 비교하는 것처럼 보여서 3스트라이크를 비교하는 거라곤 와닿진 않았던 것 같아요.😅 STRIKE_COUNT라는 상수를 살리고 싶다면 아래와 같이 판단하면 더 괜찮은 방법이겠다 싶습니다!

```javascript
function isThreeStrike(strike) {
  return strike === STRIKE_COUNT;
}

if (isThreeStrike(strikeCount)) {
  ~
}
```

nlom0218 replied:
말씀하신 내용을 생각해보니 함수로 분리해야 할 필요성이 느껴지네요. 같은 두개를 비교하는 느낌? 어색한 부분을 말씀해 주셔서 감사해요!

sangpok commented on src/View/OutputView.js:
```diff
@@ -0,0 +1,29 @@
+    const strikeCountPrint =
+      strikeCount !== 0 ? `${strikeCount}${OUTPUT.strike}` : '';
+
+    Console.print(`${ballCountPrint}${strikeCountPrint}`);
```
이렇게 출력하면 볼만 있을 때, '2볼'이 아닌 '2볼 '로 출력이 되니까, trim()으로 한 번 더 최종 출력을 정제하면 좋을 것 같습니다 !!
```suggestion
    Console.print(`${ballCountPrint}${strikeCountPrint}`.trim());
```

sangpok commented on src/libs/Quit.js:
```diff
@@ -0,0 +1,13 @@
+  },
+};
+
+module.exports = Quit;
```
저는 Quit에서 쓰이는 메소드들이 따로 분리가 되어야 한다엔 고개가 끄덕이지만 그것들이 Quit이라는 객체로 분리되는 것엔 고개가 갸우뚱거리게 되네요...! Action에 따른 분리같은데, 혹시 어떠한 생각으로 Quit이라는 객체를 떠올리셨는지 궁금합니다!🙌

nlom0218 replied:
애플리케이션이 종료되는 경우는 2가지이기 때문에 이를 묶어 `Quit`라는 객체에서 관리하고자 했어요. 혹시 주현님께서는 위와 같은 메서드를 어떻게 관리하시는 것을 추천하시나요?

sangpok replied:
> 애플리케이션이 종료되는 경우는 2가지이기 때문에 이를 묶어 `Quit`라는 객체에서 관리하고자 했어요. 혹시 주현님께서는 위와 같은 메서드를 어떻게 관리하시는 것을 추천하시나요?

저도 어디로 들어가야 할지 생각해보다가 생각이 많아졌습니다😂 일단,, 

**1. `Quit.game` 부분**
`Console.close()`는 OutputView나 InputView에 들어가는 게 맞지 않나? 하는 생각이 들었습니다. 

(1) 응집도를 생각해서, View에 필요한 초기 설정 및 마무리 단계 역시 하나의 모듈에서 이뤄지는 게 맞다.
(2) 다른 모듈에서 Console을 불러와서 close하는 작업은 모듈의 결합도가 높아진다.

라는 이유로 OutputView 또는 InputView에 들어가는 게 맞다고 생각했고, 아니면 이 두 개의 뷰를 하나로 묶는 핸들러 모듈을 만들고, 여기에 close를 다는 방법도 있다고 생각이 듭니다 :)

**2. `Quit.withException` 부분**
이것은 Validate하는 부분에서 관리를 해주는 편이 좋다는 생각입니다.  `new Error()`를 사용해서 오류 객체를 만들면 해당 줄을 기점으로 stacktrace가 생성되고, 그렇게 된다면 에러 추적이 필요할 때 추적이 힘들어진다고 배웠습니다👍

nlom0218 replied:
자세한 설명 감사합니다! 다음번에 다들 어떻게 관리하는지 확인해보고 싶네요!

sangpok commented on src/libs/const.js:
```diff
@@ -0,0 +1,35 @@
+};
+
+const ERROR_MESSAGE = {
+  isNotNumber: '[ERROR] 숫자만 입력 가능합니다.',
+  length: '[ERROR] 숫자 3개를 입력해야 합니다.',
+  includesZero: '[ERROR] 0은 입력할 수 없습니다.',
+  overlap: '[ERROR] 중복된 숫자를 입력할 수 없습니다.',
+  option: '1 또는 2를 입력하지 않아 애플리케이션이 종료됩니다.',
```
에러의 머릿말을 분리시키면 더 좋은 확장성과 명시성을 가지게 될 것 같습니다 :)

```suggestion
  errorHeading: '[ERROR]'
  isNotNumber: '숫자만 입력 가능합니다.',
  length: '숫자 3개를 입력해야 합니다.',
  includesZero: '0은 입력할 수 없습니다.',
  overlap: '중복된 숫자를 입력할 수 없습니다.',
  option: '1 또는 2를 입력하지 않아 애플리케이션이 종료됩니다.',
```

sangpok reviewed (commented):
멋진 Jest코드들과 역할 분리를 위한 객체 생성이 인상깊었습니다 :)
조금이나마 도움이 되고자 보이는 부분들을 적어보았습니다! 수고하셨습니다👏✨

nlom0218 committed:
refactor: 11월 30일 코드 리뷰에 대한 리팩터링

GameManager.#play() 중복 호출에 대한 문제 해결
bind() 메서드가 아닌 Arror Function 변경
ComputerAnswer 클래스를 재생성, View에 관련된 클래스를 객체로 변경
3스트라이크인지 판별하는 함수 생성, trim() 메서드 사용
에러의 머릿말 추가

nlom0218 committed:
refactor: else 지양하기

shackstack commented on src/Controller/GameManager.js:
```diff
@@ -0,0 +1,65 @@
+    Validator.playerAnswer(playerAnswer);
+
+    OutputView.printResult(
+      this.#computerAnswer.comparePlayerAnswer(playerAnswer)
```
별건 아니지만 compare 혹은 비교대상을 붙여서 compareAWithB 이런식으로 작명하시는 것이 더 나아보일 것 같습니다.

nlom0218 replied:
`ComputerAnswer` 클래스의 매서드이기 때문에 `comparePlayerAnswer`으로 이름을 지었어요. `computerAnswer`이 중복되는 것을 피하고 싶었습니다!ㅎㅎ 무엇이 더 좋은지 고민을 해봐야겠습니다. 항상 변수명, 메서드명을 짓는것이 까다롭더라고요ㅠㅠ

nlom0218 replied:
추가적으로 파일명, 폴더명을 통일을 하지 않은 이유는 딱히 없어요!ㅠㅠ 어떻게 보면 기준을 정해두지 않고 작성을 했기에 발생하는 문제점이라고 생각합니다.  Airbnb 코딩 컨벤션에서는 `파일 이름은 default export의 이름과 일치해야 합니다.`라고 소개가 되어 있더라고요. 파일이름은 어떻게 지어야 하는지 느낌이 오지만, 폴더명은 어떤지 궁금하네요!

shackstack commented on src/libs/const.js:
```diff
@@ -0,0 +1,37 @@
+  overlap: `${ERROR_HEADING} 중복된 숫자를 입력할 수 없습니다.`,
+  option: `${ERROR_HEADING} 1 또는 2를 입력하지 않아 애플리케이션이 종료됩니다.`,
+};
+
```
ERROR_HEADING으로 상수화시킨 부분은 좋은 것 같습니다. 그런데 이것 또한 리팩터링을 할 수 있을 것 같다는 생각이드네요. 에러메세지메이커라는 함수를 하나 만들어서 관리하면 반복되는 코드를 줄일 수 있을 것 같습니다.

nlom0218 replied:
아하! 좋은 의견 감사합니다ㅎㅎ👍

shackstack reviewed (commented):
좋은 코드 잘 배우고 갑니다. 한가지 궁금한 점이 있는데 저는 폴더명은 소문자로 통일하고 파일명은 카멜케이스로 작성합니다. 홍동님은 따로 파일명 폴더명을 통일하시진 않으신 것 같은데 이유가 있으실까요?

nlom0218 committed:
refactor: ERROR_HEADING를 불러와서 에러 메시지 작성하기
